### PR TITLE
[posix] add rcp diag command to check RCP's radio and spinel capbility flags

### PR DESCRIPTION
**Pull request by zhanglongxia**

### CODE REVIEW

File: src/posix/platform/rcp_caps_diag.cpp [hunk 1]
```diff
@@ -498,22 +502,149 @@ void RcpCapsDiag::TestSpinelCommands(Category aCategory)
...
+    Output("\r\n%s :\r\n", CategoryToString(aCategory));
+
+    for (const FlagEntry &entry : kRadioFlagEntries)
+    {
+        if (entry.mCategory != aCategory)
+        {
+            continue;
+        }
+
+        OutputFormat(RadioCapbilityToString(entry.mFlag), SupportToString((aRadioCapFlags & entry.mFlag) > 0));
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.cpp:550]:
```suggestion
        if (entry.mCategory == aCategory)
        {
            OutputFormat(RadioCapbilityToString(entry.mFlag), SupportToString((aRadioCapFlags & entry.mFlag) > 0));
        }
```

Avoid using continue as possible?

File: src/posix/platform/rcp_caps_diag.cpp [hunk 2]
```diff
@@ -498,22 +502,149 @@ void RcpCapsDiag::TestSpinelCommands(Category aCategory)
...
+    {
+        if (entry.mCategory != aCategory)
+        {
+            continue;
+        }
+
+        snprintf(capName, sizeof(capName), "SPINEL_CAPS_%s", spinel_capability_to_cstr(entry.mFlag));
+        OutputFormat(capName, SupportToString(IsSpinelCapabilitySupported(aCapsData, aCapsLength, entry.mFlag)));
+    }
+}
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.cpp:599]: What about doing like this:
```
void RcpCapsDiag::TestSpinelCapbilityFlags(FlagEntry *aFlagEntries, uint8_t aEntriesLength, const uint8_t *aCapsData, spinel_size_t aCapsLength)
{
     // Simply iterate over all flag entries and do the output.
}
```
And in `TestSpinelCapbilityFlags`:
```
void RcpCapsDiag::TestSpinelCapbilityFlags(void)
{
    static constexpr FlagEntry kSpinelFlagEntriesBasic[] = {{kCategoryBasic, SPINEL_CAP_CONFIG_RADIO},
                                                           {kCategoryBasic, SPINEL_CAP_MAC_RAW},
                                                           {kCategoryBasic, SPINEL_CAP_RCP_API_VERSION}};
    static constexpr FlagEntry kSpinelFlagEntriesUtils[] = {kCategoryUtils, SPINEL_CAP_OPENTHREAD_LOG_METADATA},
                                                           {kCategoryUtils, SPINEL_CAP_RCP_MIN_HOST_API_VERSION},
                                                           {kCategoryUtils, SPINEL_CAP_RCP_RESET_TO_BOOTLOADER}};
    ...
    TestSpinelCapbilityFlags(kSpinelFlagEntriesBasic, sizeof(kSpinelFlagEntriesBasic), ...);
    TestSpinelCapbilityFlags(kSpinelFlagEntriesUtils, sizeof(kSpinelFlagEntriesUtils), ...);

    return;
}
```
In this way, we can skip the category comparison. And we can even remove the category field in `FlagEntry`.

Just a suggestion and it's up to you.

File: src/posix/platform/rcp_caps_diag.cpp [hunk 3]
```diff
@@ -498,22 +502,149 @@ void RcpCapsDiag::TestSpinelCommands(Category aCategory)
...
+        aCapsLength -= static_cast<spinel_size_t>(unpacked);
+    }
+
+exit:
+    return ret;
+}
+
+void RcpCapsDiag::OutputFormat(const char *aName, const char *aValue)
+{
+    static constexpr uint8_t kMaxLength    = 56;
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.cpp:624]: Should this be equal to the `kMaxLength` defined in `OutputResult`? If so, what about moving these constexpr definition into the class private section?

File: src/posix/platform/rcp_caps_diag.hpp [hunk 4]
```diff
@@ -99,11 +99,26 @@ class RcpCapsDiag
...
+    void TestRadioCapbilityFlags(void);
+    void TestRadioCapbilityFlags(Category aCategory, uint32_t aRadioCapFlags);
+    void TestSpinelCapbilityFlags(void);
+    void TestSpinelCapbilityFlags(Category aCategory, const uint8_t *aCapsData, spinel_size_t aCapsLength);
+    bool IsSpinelCapabilitySupported(const uint8_t *aCapsData, spinel_size_t aCapsLength, uint32_t aCapability);
+    void OutputFormat(const char *aName, const char *aValue);
     void OutputResult(const SpinelEntry &aEntry, otError error);
     void Output(const char *aFormat, ...);
 
+    const char *SupportToString(bool aSupport);
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.hpp:120]:
```suggestion
    static const char *SupportToString(bool aSupport);
```
Declare as static?

File: src/posix/platform/rcp_caps_diag.hpp [hunk 5]
```diff
@@ -99,11 +99,26 @@ class RcpCapsDiag
...
+    const char *RadioCapbilityToString(uint32_t aCapability);
```
Comment (Irving-cl) [inline src/posix/platform/rcp_caps_diag.hpp:121]:
```suggestion
    static const char *RadioCapbilityToString(uint32_t aCapability);
```
This method can also be declared as static.

File: src/posix/platform/rcp_caps_diag.cpp (hunk 1 above)
Comment (zhanglongxia) [reply src/posix/platform/rcp_caps_diag.cpp]: Updated

File: src/posix/platform/rcp_caps_diag.cpp (hunk 2 above)
Comment (zhanglongxia) [reply src/posix/platform/rcp_caps_diag.cpp]: Updated

File: src/posix/platform/rcp_caps_diag.cpp (hunk 3 above)
Comment (zhanglongxia) [reply src/posix/platform/rcp_caps_diag.cpp]: They are two different values. I have renamed these two values to make the definition clearer.

**Review by Irving-cl — APPROVED** [review]
LGTM 👍

File: src/posix/platform/rcp_caps_diag.cpp [hunk 6]
```diff
@@ -547,6 +675,38 @@ const char *RcpCapsDiag::CategoryToString(Category aCategory)
...
+    static_assert(OT_RADIO_CAPS_ENERGY_SCAN == 1 << 1, "OT_RADIO_CAPS_ENERGY_SCAN value is incorrect");
+    static_assert(OT_RADIO_CAPS_TRANSMIT_RETRIES == 1 << 2, "OT_RADIO_CAPS_TRANSMIT_RETRIES value is incorrect");
+    static_assert(OT_RADIO_CAPS_CSMA_BACKOFF == 1 << 3, "OT_RADIO_CAPS_CSMA_BACKOFF value is incorrect");
+    static_assert(OT_RADIO_CAPS_SLEEP_TO_TX == 1 << 4, "OT_RADIO_CAPS_SLEEP_TO_TX value is incorrect");
+    static_assert(OT_RADIO_CAPS_TRANSMIT_SEC == 1 << 5, "OT_RADIO_CAPS_TRANSMIT_SEC value is incorrect");
+    static_assert(OT_RADIO_CAPS_TRANSMIT_TIMING == 1 << 6, "OT_RADIO_CAPS_TRANSMIT_TIMING value is incorrect");
+    static_assert(OT_RADIO_CAPS_RECEIVE_TIMING == 1 << 7, "OT_RADIO_CAPS_RECEIVE_TIMING value is incorrect");
+    static_assert(OT_RADIO_CAPS_RX_ON_WHEN_IDLE == 1 << 8, "OT_RADIO_CAPS_RX_ON_WHEN_IDLE value is incorrect");
+
+    int32_t pos = static_cast<int32_t>(ffs(static_cast<int>(aCapability))) - 1;
```
Comment (abtink) [inline src/posix/platform/rcp_caps_diag.cpp:704]: Suggest maybe not using the `ffs()` here (avoid relying on library support for this). 
We can consider implementing our own version of such a function?

Comment (abtink) [reply src/posix/platform/rcp_caps_diag.cpp:704]: How about something like this:
```c++
    const char *string = "invalid;"
    uint16_t   index   = 0;

    for (; !(aCapability & 0x1); aCapability >>= 1, index++)
    {
        VerifyOrExit(index < OT_ARRAY_LENGTH(kCapbilityStrings));
    }

    string = kCapbilityStrings[index];

exit:
    return string;
```

**Review by abtink — COMMENTED** [review]
Looks good. Thanks.
One suggestion below.

File: src/posix/platform/rcp_caps_diag.cpp (hunk 6 above)
Comment (zhanglongxia) [reply src/posix/platform/rcp_caps_diag.cpp]: Updated

File: src/posix/platform/README_RCP_CAPS_DIAG.md [hunk 7]
```diff
@@ -19,14 +58,57 @@ Check which Spinel commands RCP supports.
...
+PROP_VALUE_GET RADIO_COEX_ENABLE -------------------------- OK
+PROP_VALUE_GET RADIO_COEX_METRICS ------------------------- OK
+PROP_VALUE_GET RCP_MIN_HOST_API_VERSION ------------------- OK
+PROP_VALUE_SET PHY_CCA_THRESHOLD -------------------------- OK
+PROP_VALUE_SET PHY_CHAN_MAX_POWER ------------------------- OK
+PROP_VALUE_SET PHY_CHAN_TARGET_POWER ---------------------- OK
+PROP_VALUE_SET PHY_FEM_LNA_GAIN --------------------------- OK
+PROP_VALUE_SET PHY_REGION_CODE ---------------------------- OK
+PROP_VALUE_SET PHY_TX_POWER ------------------------------- OK
+PROP_VALUE_SET RADIO_COEX_ENABLE -------------------------- OK
```
Comment (abtink) [inline src/posix/platform/README_RCP_CAPS_DIAG.md:112]: The output format looks really neat. 👍

**Review by abtink — APPROVED** [review]
Looks great. Thanks. 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**